src/h5web/explorer/utils.ts

```diff
@@ -9,83 +9,110 @@ import {
   MyHDF5Datatype,
   HDF5HardLink,
   MyHDF5Metadata,
+  MyHDF5EntityKind,
+  HDF5Datatype,
+  HDF5Dataset,
 } from '../providers/models';
-import {
-  makeMyDataset,
-  makeMyDatatype,
-  makeMyGroup,
-  makeMyLink,
-} from '../providers/my-utils';
-import { isGroup, isReachable } from '../providers/utils';
+import { isGroup, isReachable } from '../guards';
 import { getChildEntity } from '../visualizations/nexus/utils';
+import { nanoid } from 'nanoid';
 
-function buildDataset(
-  metadata: HDF5Metadata,
-  link: HDF5HardLink
-): MyHDF5Dataset {
-  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-  const { shape, type, attributes } = metadata.datasets![link.id];
+function buildDataset(dataset: HDF5Dataset, link: HDF5HardLink): MyHDF5Dataset {
+  const { shape, type, attributes = [] } = dataset;
 
-  return makeMyDataset(link.title, shape, type, {
+  return {
+    uid: nanoid(),
     id: link.id,
+    name: link.title,
+    kind: MyHDF5EntityKind.Dataset,
     attributes,
+    shape,
+    type,
     rawLink: link,
-  });
+  };
 }
 
 function buildDatatype(
-  metadata: HDF5Metadata,
+  datatype: HDF5Datatype,
   link: HDF5HardLink
 ): MyHDF5Datatype {
-  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-  const { type } = metadata.datatypes![link.id];
+  const { type } = datatype;
 
-  return makeMyDatatype(link.title, type, {
+  return {
+    uid: nanoid(),
     id: link.id,
+    name: link.title,
+    kind: MyHDF5EntityKind.Datatype,
+    attributes: [],
+    type,
     rawLink: link,
-  });
+  };
 }
 
 function buildGroup(
-  metadata: HDF5Metadata,
+  metadata: Required<HDF5Metadata>,
   link: HDF5HardLink | HDF5RootLink
 ): MyHDF5Group {
   const rawGroup = metadata.groups[link.id];
 
   const children = (rawGroup.links || []).map((link) => {
     if (!isReachable(link)) {
-      return makeMyLink(link);
+      return {
+        uid: nanoid(),
+        name: link.title,
+        kind: MyHDF5EntityKind.Link,
+        attributes: [],
+        rawLink: link,
+      };
     }
 
     switch (link.collection) {
       case HDF5Collection.Groups:
         return buildGroup(metadata, link);
       case HDF5Collection.Datasets:
-        return buildDataset(metadata, link);
+        return buildDataset(metadata.datasets[link.id], link);
       case HDF5Collection.Datatypes:
-        return buildDatatype(metadata, link);
+        return buildDatatype(metadata.datatypes[link.id], link);
```

Comment:
Passing the raw entities instead of the whole metadata object just simplifies the signature of `buildDataset` and `buildDatatype`.